engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbDecisionState.java

```diff
@@ -209,6 +209,38 @@ public void storeDecisionRequirements(final DecisionRequirementsRecord record) {
     updateLatestDecisionRequirementsVersion(record);
   }
 
+  @Override
+  public void deleteDecision(final DecisionRecord record) {
+    findLatestDecisionById(record.getDecisionIdBuffer())
+        .map(PersistedDecision::getVersion)
+        .ifPresent(
+            latestVersion -> {
+              if (latestVersion == record.getVersion()) {
+                dbDecisionId.wrapBuffer(record.getDecisionIdBuffer());
+                findPreviousVersionDecisionKey(record.getDecisionIdBuffer(), record.getVersion())
+                    .ifPresentOrElse(
+                        previousDecisionKey -> {
+                          // Update the latest decision version
+                          dbDecisionKey.wrapLong(previousDecisionKey);
+                          latestDecisionKeysByDecisionId.update(dbDecisionId, fkDecision);
+                        },
+                        () -> {
+                          // Clear the latest decision version
+                          latestDecisionKeysByDecisionId.deleteExisting(dbDecisionId);
+                        });
```

Comment:
There are a few changes you could try to make this method a bit easier to  follow, e.g. reduce the nesting levels.

🔧 This if statement could be moved up into a filter function, reducing the indentation of the rest of the lambda. You could even combine the `map` and `filter` because what you're effectively trying to achieve is just filtering the optional. You don't really care about the 'found version' because if it matches you already knew it as 'record.version'

🔧 Since you don't use `latestVersion` anymore in these nested lambdas of `isPresentOrElse`, you could change the `ifPresent` above to `flatMap` (returning the result of the call to `findPreviousVersionDecisionKey`). This allows you to move the `ifPresentOrElse` call to the end of the top-most call chain.

```java
//pseudo code
findLatest()
  .filter(latest -> latest.version == record.version)
  .flatMap(findPreviousVersion)
  .ifPresentOrElse(a, b)
```

Reply:
I understand what you're saying, but there is slight behavioural difference int he way I build it, and the way you describe.

The `filter` will return an empty optional if the predicate does not match. As a result `b` in the `ifPresentOrElse` will get executed if the latest version does not match the record version. This is not what we want to do. We only want to perform a or b if the latest version matches.